File: java/DataStructureBenchmark/src/structures/Graph/Graph.java
package structures.Graph;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import structures.common.Items;

public class Graph {
	// 해시맵 기반의 인접 리스트로 구현
	private final Map<String, Items> vertices = new HashMap<>(); // 각 정점의 키와 관련된 데이터를 저장하는 해시맵
	private final Map<String, List<String>> adjList = new HashMap<>(); // 각 정점의 키와 인접한 정점들의 리스트를 저장하는 해시맵(인접 리스트의 변형 형태)
	private final Map<String, String> lastByPrefix = new HashMap<>(); // 각 접두사에 대해 마지막으로 삽입된 정점의 키를 저장하는 해시맵

	// insert
	public void insert(String key, String keyString, String value) {
		vertices.put(key, new Items(key, keyString, value));
		adjList.putIfAbsent(key, new ArrayList<>());

		String prefix = keyString.contains("-") ? keyString.split("-")[0] : "default";

		if (lastByPrefix.containsKey(prefix)) {
			String prevKey = lastByPrefix.get(prefix);
			adjList.get(prevKey).add(key);
			adjList.get(key).add(prevKey);
		}

		lastByPrefix.put(prefix, key);
	}

	// search	
	public Items search(String key) {
		return vertices.get(key);
	}

}

File: php/DataStructureBenchmark/structures/Graph.php
<?php

declare(strict_types=1);

class Graph
{
    private array $vertices = []; // key => ['keyString'=>..., 'value'=>...]
    private array $adjList = [];  // key => [neighborKey1, neighborKey2, ...]
    private array $lastByPrefix = [];

    public function insert(string $key, string $keyString, string $value): void
    {
        $this->vertices[$key] = [
            'key' => $key,
            'keyString' => $keyString,
            'value' => $value,
        ];

        if (!isset($this->adjList[$key])) {
            $this->adjList[$key] = [];
        }

        $prefix = explode('-', $keyString)[0] ?? 'default';

        if (isset($this->lastByPrefix[$prefix])) {
            $prevKey = $this->lastByPrefix[$prefix];
            $this->adjList[$prevKey][] = $key;
            $this->adjList[$key][] = $prevKey;
        }

        $this->lastByPrefix[$prefix] = $key;
    }

    public function search(string $key): ?array
    {
        return $this->vertices[$key] ?? null;
    }

    public function sortByKeyString(): void
    {
        uasort($this->vertices, static function (array $a, array $b): int {
            return strcmp($a['keyString'], $b['keyString']);
        });
    }

    public function count(): int
    {
        return count($this->vertices);
    }
}